File: packages/sdk-ts/src/manifest/offsets.ts
/**
 * Offset arithmetic for prompt manifests.
 *
 * Manifest `charStart` / `charEnd` are Unicode CODE-POINT indices into
 * the source file content. Not UTF-16 code units (JS / TS native) and
 * not UTF-8 bytes (Go native). Code points are the only unit that
 * matches across all three SDK languages without conversion math at
 * every read/write boundary — Python's `str[a:b]` is already code
 * points, so picking code points lets the SDK handler slice manifest
 * entries directly.
 *
 * Use these helpers wherever a manifest offset is read or written.
 * Never pass a manifest offset to plain `string.slice()` — that uses
 * UTF-16 code units and will split surrogate pairs (any astral char,
 * including 🎯 and most emoji, takes two code units).
 *
 * The Go counterparts live in `cli/internal/manifest/offsets.go`.
 */

/** Number of Unicode code points in `s`. */
export function codePointLength(s: string): number {
  let n = 0
  // for…of iterates by code point, surrogate-pair safe.
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  for (const _ of s) n++
  return n
}

/**
 * Substring of `s` from code-point index `cpStart` (inclusive) to
 * `cpEnd` (exclusive). Out-of-range indices clamp to
 * `[0, codePointLength(s)]`. Empty string for empty / inverted ranges.
 */
export function sliceByCodePoints(s: string, cpStart: number, cpEnd: number): string {
  if (cpStart < 0) cpStart = 0
  if (cpEnd < cpStart) cpEnd = cpStart
  let cp = 0
  let unitOff = 0
  let unitStart = -1
  let unitEnd = -1
  for (const ch of s) {
    if (cp === cpStart && unitStart < 0) unitStart = unitOff
    if (cp === cpEnd) {
      unitEnd = unitOff
      break
    }
    unitOff += ch.length // 1 for BMP, 2 for surrogate pair
    cp++
  }
  if (unitStart < 0) {
    // cpStart at or past end of string.
    if (cpStart === cp) return ''
    return ''
  }
  if (unitEnd < 0) unitEnd = s.length
  return s.slice(unitStart, unitEnd)
}

/**
 * Convert a UTF-16 code-unit index (what JS strings natively use) to
 * a Unicode code-point index. Code units that land mid-surrogate-pair
 * snap to the next code point.
 */
export function codeUnitToCodePoint(s: string, unitIdx: number): number {
  if (unitIdx <= 0) return 0
  let cp = 0
  let off = 0
  for (const ch of s) {
    if (off >= unitIdx) return cp
    off += ch.length
    cp++
  }
  return cp
}

/** Convert a code-point index to its UTF-16 code-unit index. */
export function codePointToCodeUnit(s: string, cp: number): number {
  if (cp <= 0) return 0
  let n = 0
  let off = 0
  for (const ch of s) {
    if (n === cp) return off
    off += ch.length
    n++
  }
  return s.length
}

/**
 * Code-point offset of the start of the (0-indexed) line N. `line == 0`
 * returns 0. Past EOF returns -1; the "trailing line with no \n" case
 * returns `codePointLength(s)` for `line == lineCount`.
 */
export function lineToCodePointOffset(s: string, line: number): number {
  if (line < 0) return -1
  if (line === 0) return 0
  let cp = 0
  let lc = 0
  for (const ch of s) {
    if (ch === '\n') {
      cp++
      lc++
      if (lc === line) return cp
      continue
    }
    cp++
  }
  if (lc === line - 1) return cp
  if (lc >= line) return cp
  return -1
}

/**
 * Code-point [start, end) bounds of (1-indexed) line N — `end` is one
 * past the line's last non-newline character. Returns `[-1, -1]` past
 * EOF or for non-positive `line`. Use this to bound an anchored
 * substring search to a single line.
 */
export function lineContentCodePoints(s: string, line: number): [number, number] {
  if (line < 1) return [-1, -1]
  const start = lineToCodePointOffset(s, line - 1)
  if (start < 0) return [-1, -1]
  const nextLineStart = lineToCodePointOffset(s, line)
  if (nextLineStart < 0) return [-1, -1]
  let end = nextLineStart
  if (end > start && sliceByCodePoints(s, end - 1, end) === '\n') end--
  return [start, end]
}

File: cli/internal/manifest/offsets.go
package manifest

import "unicode/utf8"

// Offset arithmetic for prompt manifests.
//
// Manifest `charStart` / `charEnd` are Unicode CODE-POINT indices into
// the source file content. Not bytes (Go's native string indexing) and
// not UTF-16 code units (JS / TS native). Code points are the only unit
// that matches across Go, TS, and Python without conversion math at
// every read/write boundary; Python's `str[a:b]` is already code points,
// so picking code points lets the SDK handler slice manifest entries
// directly.
//
// Use these helpers wherever a manifest offset is read or written. Never
// pass a manifest offset to a Go byte-slice or a TS `string.slice()`
// without going through SliceByCodePoints / sliceByCodePoints first —
// any non-ASCII character (em-dash, smart quote, emoji, accented letter)
// will desync the indices and chop the wrong characters.

// CodePointLen returns the number of Unicode code points in s.
func CodePointLen(s string) int {
	return utf8.RuneCountInString(s)
}

// SliceByCodePoints returns the substring of s from code-point index
// cpStart (inclusive) to cpEnd (exclusive). Out-of-range indices clamp
// to [0, CodePointLen(s)]. Returns "" when the clamped range is empty.
func SliceByCodePoints(s string, cpStart, cpEnd int) string {
	if cpStart < 0 {
		cpStart = 0
	}
	if cpEnd < cpStart {
		cpEnd = cpStart
	}
	cp := 0
	byteStart, byteEnd := -1, -1
	for i := range s { // range over a string iterates by rune; i is the byte offset
		if cp == cpStart && byteStart < 0 {
			byteStart = i
		}
		if cp == cpEnd {
			byteEnd = i
			break
		}
		cp++
	}
	if byteStart < 0 {
		// cpStart is at or past the end of the string.
		if cpStart == cp {
			return ""
		}
		return ""
	}
	if byteEnd < 0 {
		// cpEnd ran past the end; clamp to len(s).
		byteEnd = len(s)
	}
	return s[byteStart:byteEnd]
}

// CodePointToByteOffset converts a code-point index to its UTF-8 byte
// offset. Out-of-range inputs clamp to 0 / len(s).
func CodePointToByteOffset(s string, cp int) int {
	if cp <= 0 {
		return 0
	}
	n := 0
	for i := range s {
		if n == cp {
			return i
		}
		n++
	}
	return len(s)
}

// ByteOffsetToCodePoint converts a UTF-8 byte offset to the code-point
// index of the character starting at that byte. byteOff must be on a
// UTF-8 boundary; values that land mid-rune snap to the next rune.
func ByteOffsetToCodePoint(s string, byteOff int) int {
	if byteOff <= 0 {
		return 0
	}
	cp := 0
	for i := range s {
		if i >= byteOff {
			return cp
		}
		cp++
	}
	return cp
}

// LineToCodePointOffset returns the code-point offset (into s) of the
// start of the (0-indexed) line N. line == 0 returns 0 (start of file).
// line >= total-line-count returns CodePointLen(s) when s has no
// trailing newline on its last line; -1 only for negative input or a
// requested line strictly past EOF.
//
// Newline character (\n) counts as one code point on the line it ends.
// CRLF is not recognised; callers that need it should strip \r first.
func LineToCodePointOffset(s string, line int) int {
	if line < 0 {
		return -1
	}
	if line == 0 {
		return 0
	}
	cp := 0
	lineCount := 0
	for _, r := range s {
		if r == '\n' {
			cp++
			lineCount++
			if lineCount == line {
				return cp
			}
			continue
		}
		cp++
	}
	// Past the final newline: the requested line is the trailing
	// (no-newline) line. Its start is end-of-file.
	if lineCount == line-1 {
		return cp
	}
	if lineCount >= line {
		return cp
	}
	return -1
}

// LineContentCodePoints returns the code-point start and end offsets of
// (1-indexed) line N in s — `end` points one past the line's last
// non-newline character. Returns (-1, -1) if `line` is past EOF or
// non-positive. Use this to bound an anchored substring search to a
// single line.
func LineContentCodePoints(s string, line int) (start, end int) {
	if line < 1 {
		return -1, -1
	}
	start = LineToCodePointOffset(s, line-1)
	if start < 0 {
		return -1, -1
	}
	nextLineStart := LineToCodePointOffset(s, line)
	if nextLineStart < 0 {
		// Past EOF — line doesn't exist.
		return -1, -1
	}
	end = nextLineStart
	// Drop the trailing \n that ended this line, if any.
	if end > start && SliceByCodePoints(s, end-1, end) == "\n" {
		end--
	}
	return start, end
}
